lib/irb/nesting_parser.rb

```diff
@@ -0,0 +1,205 @@
+module IRB
+  module NestingParser
+    IGNORE_TOKENS = %i[on_sp on_ignored_nl on_comment on_embdoc_beg on_embdoc on_embdoc_end]
+    def self.scan(tokens)
```

Comment:
From its callers, it seems to return `opens`? Which I assume is an array of tokens that's "unclosed", like `(`?

If it is, I have some suggestions:

- Can you write some comments above the method as it's not super easy to understand due to its complexity
- Can we name it more explicitly, like `scan_opens`?

Reply:
I changed it to `scan_opens`. I also add `open_tokens` method that doesn't receive block and just return open tokens because `scan_opens` with block is only used inside nesting_parser.